File: vane-core/src/main/java/org/oddlama/vane/core/command/check/ExecutorCheckResult.java
package org.oddlama.vane.core.command.check;

import java.util.ArrayList;
import org.bukkit.command.CommandSender;
import org.oddlama.vane.core.command.Command;
import org.oddlama.vane.core.command.Executor;

public class ExecutorCheckResult implements CheckResult {

    private int depth;
    private Executor executor;
    private ArrayList<Object> parsedArgs = new ArrayList<>();

    public ExecutorCheckResult(int depth, Executor executor) {
        this.depth = depth;
        this.executor = executor;
    }

    @Override
    public int depth() {
        return depth;
    }

    @Override
    public boolean good() {
        return true;
    }

    public boolean apply(Command<?> command, CommandSender sender) {
        return executor.execute(command, sender, parsedArgs);
    }

    @Override
    public CheckResult prepend(String argumentType, Object parsedArg, boolean include) {
        if (include) {
            parsedArgs.add(0, parsedArg);
        }
        return this;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/command/check/ExecutorCheckResult.kt
package org.oddlama.vane.core.command.check

import org.bukkit.command.CommandSender
import org.oddlama.vane.core.command.Command
import org.oddlama.vane.core.command.Executor

/**
 * Successful match result that holds an executable callback and parsed arguments.
 *
 * @param depth parse depth where this executor matched.
 * @param executor command executor callback.
 */
class ExecutorCheckResult(private val depth: Int, private val executor: Executor) : CheckResult {
    /**
     * Parsed arguments collected while prepending match results.
     */
    private val parsedArgs = mutableListOf<Any?>()

    /** Returns the parse depth for this result. */
    override fun depth(): Int = depth

    /** Executor results are always successful. */
    override fun good(): Boolean = true

    /** Executes the stored callback with collected arguments. */
    override fun apply(command: Command<*>?, sender: CommandSender?): Boolean =
        executor.execute(command, sender, parsedArgs)

    /** Prepends a parsed argument into the execution argument list when requested. */
    override fun prepend(argumentType: String, parsedArg: Any?, include: Boolean): CheckResult {
        if (include) parsedArgs.add(0, parsedArg)
        return this
    }
}
